allennlp/semparse/worlds/world.py

```diff
@@ -0,0 +1,519 @@
+                               "Are you sure you want a unary production rule in your grammar?")
+        return remaining_actions
+
+    @classmethod
+    def _infer_num_arguments(cls, type_signature: str) -> int:
+        """
+        Takes a type signature and infers the number of arguments the corresponding function takes.
+        Examples:
+            e -> 0
+            <r,e> -> 1
+            <e,<e,t>> -> 2
+            <b,<<b,#1>,<#1,b>>> -> 3
+        """
+        if not "<" in type_signature:
+            return 0
+        # We need to find the return type from the signature. We do that by removing the outer most
+        # angular brackets and traversing the remaining substring till the angular brackets (if any)
+        # balance. Once we hit a comma after the angular brackets are balanced, whatever is left
+        # after it is the return type.
+        type_signature = type_signature[1:-1]
+        num_brackets = 0
+        char_index = 0
+        for char in type_signature:
+            if char == '<':
+                num_brackets += 1
+            elif char == '>':
+                num_brackets -= 1
+            elif char == ',':
+                if num_brackets == 0:
+                    break
+            char_index += 1
+        return_type = type_signature[char_index+1:]
+        return 1 + cls._infer_num_arguments(return_type)
+
+    def _process_nested_expression(self, nested_expression) -> str:
+        """
+        ``nested_expression`` is the result of parsing a logical form in Lisp format.
+        We process it recursively and return a string in the format that NLTK's ``LogicParser``
+        would understand.
+        """
+        expression_is_list = isinstance(nested_expression, list)
+        expression_size = len(nested_expression)
+        if expression_is_list and expression_size == 1 and isinstance(nested_expression[0], list):
+            return self._process_nested_expression(nested_expression[0])
+        elements_are_leaves = [isinstance(element, str) for element in nested_expression]
+        if all(elements_are_leaves):
+            mapped_names = [self._map_name(name) for name in nested_expression]
+        else:
+            mapped_names = []
+            for element, is_leaf in zip(nested_expression, elements_are_leaves):
+                if is_leaf:
+                    mapped_names.append(self._map_name(element))
+                else:
+                    mapped_names.append(self._process_nested_expression(element))
+        if mapped_names[0] == "\\":
+            # This means the predicate is lambda. NLTK wants the variable name to not be within parantheses.
+            # Adding parentheses after the variable.
+            arguments = [mapped_names[1]] + ["(%s)" % name for name in mapped_names[2:]]
+        else:
+            arguments = ["(%s)" % name for name in mapped_names[1:]]
+        return "(%s %s)" % (mapped_names[0], " ".join(arguments))
```

Comment:
python 3.6 strings, Matt I thought I knew you

Reply:
Done.  Looks like I missed this when I first reviewed Pradeep's code =).